easybuild/framework/easyblock.py

```diff
@@ -3350,6 +3353,332 @@ def _sanity_check_step_multi_deps(self, *args, **kwargs):
         self.cfg['builddependencies'] = builddeps
         self.cfg.iterating = False
 
+    def sanity_check_cuda(self, cuda_dirs=None, check_cuobjdump=True):
+        """Sanity check that binaries/libraries contain device code for the correct architecture targets."""
+
+        self.log.info("Checking binaries/libraries for CUDA device code...")
+
+        fail_msgs = []
+        cfg_ccs = build_option('cuda_compute_capabilities') or self.cfg.get('cuda_compute_capabilities', None)
+        ignore_failures = not build_option('cuda_sanity_check_error_on_fail')
+        strict_cc_check = build_option('cuda_sanity_check_strict')
+        accept_ptx_as_devcode = build_option('cuda_sanity_check_accept_ptx_as_devcode')
+        accept_missing_ptx = build_option('cuda_sanity_check_accept_missing_ptx')
+
+        # Construct the list of files to ignore as full paths (cuda_sanity_ignore_files contains the paths
+        # to ignore, relative to the installation prefix)
+        ignore_file_list = [os.path.join(self.installdir, d) for d in self.cfg['cuda_sanity_ignore_files']]
+
+        # If there are no CUDA compute capabilities defined, return
+        if cfg_ccs is None or len(cfg_ccs) == 0:
+            self.log.info("Skipping CUDA sanity check, as no CUDA compute capabilities were configured")
+            return fail_msgs
+
+        if cuda_dirs is None:
+            cuda_dirs = self.cfg['bin_lib_subdirs'] or self.bin_lib_subdirs()
+
+        if not cuda_dirs:
+            cuda_dirs = DEFAULT_BIN_LIB_SUBDIRS
+            self.log.info("Using default subdirectories for binaries/libraries to verify CUDA device code: %s",
+                          cuda_dirs)
+        else:
+            self.log.info("Using configured subdirectories for binaries/libraries to verify CUDA device code: %s",
+                          cuda_dirs)
+
+        # Tracking number of CUDA files for a summary report:
+        num_cuda_files = 0
+
+        # Creating lists of files for summary report:
+        files_missing_devcode = []
+        files_missing_devcode_fails = []
+        files_missing_devcode_ignored = []
+        files_additional_devcode = []
+        files_additional_devcode_fails = []
+        files_additional_devcode_ignored = []
+        files_missing_ptx = []
+        files_missing_ptx_fails = []
+        files_missing_ptx_ignored = []
+        files_missing_devcode_but_has_ptx = []
```

Comment:
When it comes to printing these later, can we format them so they are printed in a more easy-to-read way in the logs later? Something like
```
files_missing_ptx_str = "\n" + "\n".join(f"  {f}" for f in files_missing_ptx)
```

Reply:
You mean: in addition to having the files on a new line after the 
"XXX files missing PTX code for the highest configured CUDA Compute Capability" that we don't print a single _line_ listing the files, but _each_ file on a single line? Sure, that's possible. But note that it might be a _lot_ of files, and thus a lot of lines! But you may be right that a single _very_ long line is probably still worse ;-)

Reply:
It might just be quicker than adding all those `\n` later as per my original suggestion, preformatting the lists should make that a bit easier. Could be done via a function so we can easily tweak this later:
```
def format_file_list_as_string(files_list):
    return "\n" + "\n".join(f"  {f}" for f in files_list)
``

Reply:
Done in https://github.com/easybuilders/easybuild-framework/pull/4692/commits/2442d4491b1a9c224bdbbee63bc5a0d2ac605598